### PR TITLE
Fix model loading time through prefetching the file on another thread

### PR DESCRIPTION
**Pull request by CoderRC**
I found that when using the current llama.cpp that the first model loading time went down due to OS not knowing which location of the file will be read and it caused a 4kb requests to disk, by reading the file on the seperate thread it will now cause only 1mb requests. Which for example 1000000 4kb requests to disk is decreased down to 4000 1mb requests to disk which is a big difference.
Specifically solves https://github.com/ggerganov/llama.cpp/issues/705
Right now it read for me 500 mb/s, after this patch it now reads 2.3 gb/s.

Try: git clone https://github.com/CoderRC/llama.cpp
to see the difference of first loading time only after restart to produce accurate results

From:
llama_print_timings:        load time = 31928.12 ms

To:
llama_print_timings:        load time = 11478.47 ms

### CODE REVIEW

**Comment by danielzgtg** [comment]
It would be better if we use `madvise` and `MADV_WILLNEED` on Linux and the equivalent on Windows. This way the OS will still read the pages from the disk without needing to copy them to the separate thread.

**Comment by comex** [comment]
Very interesting that Windows is not issuing correctly sized reads by default.  Which tool are you using to measure this?

That said, this solution is really hacky.  You probably want to use [`PrefetchVirtualMemory`](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-prefetchvirtualmemory) on Windows and madvise on Unix.

**Comment by CoderRC** [comment]
I just on windows empty the standby list through RAM MAP utility in windows and then run the llama.cpp and go to taskmgr and look at the performance disk tab.

File: ggml.h [hunk 1]
```diff
@@ -773,6 +773,19 @@ int ggml_cpu_has_blas(void);
...
+//
+#ifndef _POSIX_THREADS
+#if defined(_WIN32)
+#include <windows.h>
+#endif
+typedef HANDLE pthread_t;
+typedef DWORD thread_ret_t;
+static int pthread_create(pthread_t* out, void* unused, thread_ret_t(*func)(void*), void* arg);
+static int pthread_join(pthread_t thread, void* unused);
+#endif
```
Comment (danielzgtg) [inline ggml.h:787]: Why reinvent `std::thread`?

Comment (CoderRC) [reply ggml.h:787]: std::thread does not exist in msvc

**Comment by CoderRC** [comment]
Gonna try PrefetchVirtualMemory before draft to non draft.

File: ggml.h (hunk 1 above)
Comment (danielzgtg) [reply ggml.h:787]: Seems to exist at https://learn.microsoft.com/en-us/cpp/standard-library/thread-class?view=msvc-170  . But I didn't test, so feel free to resolve this thread if `std::thread` doesn't work on your machine

**Comment by CoderRC** [comment]
PrefetchVirtualMemory is perfect and it works on windows properly. Gonna do some POSIX tests and update https://github.com/CoderRC/libmingw32_extended

**Comment by prusnak** [comment]
I don't like the fact that this approach complicates stuff also for platforms that are not broken (i.e. not Windows).

Can we get rid of everything else and keep just the PrefetchVirtualMemory call?

In another word, does this fix the issue with no further changes required?

``` patch
diff --git a/llama.cpp b/llama.cpp
index 854bb89..78fbace 100644
--- a/llama.cpp
+++ b/llama.cpp
@@ -327,6 +327,11 @@ static void *mmap_file(const char *fname, uint64_t *mm_length) {
     void *addr = MapViewOfFile(hMapping, FILE_MAP_READ, 0, 0, 0);
     CloseHandle(hMapping);
     if (!addr) return 0;
+    // Prefetch the virtual memory range
+    WIN32_MEMORY_RANGE_ENTRY range;
+    range.VirtualAddress = addr;
+    range.NumberOfBytes = (SIZE_T)length;
+    PrefetchVirtualMemory(GetCurrentProcess(), 1, &range, 0);
 #else
     int fd = open(fname, O_RDONLY);
     if (fd == -1) return 0;
```

**Comment by danielzgtg** [comment]
> not broken (i.e. not Windows).

madvise should speed things up for Linux too. It might not be as dramatic, but there should be a tiny improvement

**Comment by prusnak** [comment]
> madvise should speed things up for Linux too. It might not be as dramatic, but there should be a tiny improvement

Came up with https://github.com/ggerganov/llama.cpp/pull/740 - it speeds the loading time of 7B model 3.5 times on my macbook

**Comment by CoderRC** [comment]
Try on linux it works and compare speeds to https://github.com/ggerganov/llama.cpp/pull/740

**Comment by prusnak** [comment]
You changed end-of-line format (from Unix to Windows), so GitHub shows you replaced all the lines. Please fix this. Also don't merge main into your branch. What you should do is to rebase: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

**Comment by CoderRC** [comment]
Try to do tests and tell result below and compare to https://github.com/ggerganov/llama.cpp/pull/740

**Comment by comex** [comment]
@CoderRC Can you please test #740 as well?

**Comment by x02Sylvie** [comment]
With this PR my HDD's loading speed went from 5 mb/s to 50 mb/s of model

**Comment by CoderRC** [comment]
I am going to close this since https://github.com/ggerganov/llama.cpp/pull/801 was merged and this merge solved my model loading time for msys2 with mingw32. If want to solve Efficient preloading for mmap() go to https://github.com/ggerganov/llama.cpp/pull/869

**Pull request closed without merging**